baseapp/queryrouter.go

```diff
@@ -37,3 +49,64 @@ func (qrt *QueryRouter) AddRoute(path string, q sdk.Querier) sdk.QueryRouter {
 func (qrt *QueryRouter) Route(path string) sdk.Querier {
 	return qrt.routes[path]
 }
+
+func (qrt *QueryRouter) RegisterService(sd *grpc.ServiceDesc, handler interface{}) {
+	// adds a top-level querier based on the GRPC service name
+	qrt.routes[sd.ServiceName] =
+		func(ctx sdk.Context, path []string, req abci.RequestQuery) ([]byte, error) {
+			path0 := path[0]
+			for _, md := range sd.Methods {
+				// checks each GRPC service method to see if it matches the path
+				if md.MethodName == path0 {
+					res, err := md.Handler(handler, sdk.WrapSDKContext(ctx), func(i interface{}) error {
+						return protoCodec.Unmarshal(req.Data, i)
+					}, nil)
+					if err != nil {
+						return nil, err
+					}
+					return protoCodec.Marshal(res)
+				}
+			}
+			return nil, sdkerrors.Wrapf(sdkerrors.ErrUnknownRequest, "unknown query path: %s", path[0])
+		}
+}
+
+// QueryServiceTestHelper provides a helper for making grpc query service
+// rpc calls in unit tests. It implements both the grpc Server and ClientConn
+// interfaces needed to register a query service server and create a query
+// service client.
+type QueryServiceTestHelper struct {
+	*QueryRouter
+	ctx sdk.Context
+}
+
+func NewQueryServerTestHelper(ctx sdk.Context) *QueryServiceTestHelper {
+	return &QueryServiceTestHelper{QueryRouter: NewQueryRouter(), ctx: ctx}
+}
+
+func (q *QueryServiceTestHelper) Invoke(ctx gocontext.Context, method string, args, reply interface{}, opts ...grpc.CallOption) error {
```

Comment:
The CLIContext QueryConn actually targets a node. This just targets a query router during tests

Reply:
I indicated in the godoc comment that this implements the grpc ClientConn.Invoke method which I think is most accurate because this is not meant to be called directly but rather by generated grpc methods.